File: leetcode-algorithms/029. Divide Two Integers/solution.py
def divide(dividend, divisor):
	"""
	:type dividend: int
	:type divisor: int
	:rtype: int
	"""
	if not divisor or (dividend== -2147483648 and divisor==-1):
		return 2147483647
	sign = -1 if ((dividend<0)^(divisor<0)) else 1
	dvd = abs(dividend)
	dvs = abs(divisor)
	res = 0
	while dvd>=dvs:
		tmp = dvs
		multiple = 1
		while dvd>=(tmp<<1):
			tmp <<= 1
			multiple <<= 1
		dvd -= tmp
		res += multiple
	return res if sign==1 else -res

if __name__ == '__main__':
	dividend = 15
	divisor =3
	res = divide(dividend, divisor)
	print(res)


	print(2<<30)

File: leetcode-algorithms/029. Divide Two Integers/solution.cpp
#include <iostream>
using namespace std;
int divide(int dividend, int divisor) {
	if(divisor==0 || (dividend == INT_MIN && divisor==-1))
		return INT_MAX;

	int sign = ((divisor<0)^(dividend<0))?-1:1;
	long long dvd = labs(dividend);
	long long dvs = labs(divisor);
	int res = 0;
	while(dvd>=dvs)
	{
		long long tmp = dvs;
		int mulpiple = 1;
		while(dvd>=(tmp<<1))
		{
			tmp <<= 1;
			mulpiple <<= 1;
		}
		dvd -= tmp;
		res += mulpiple;
	}
	return sign==1?res:-res;
}
int main()
{
	int dividend = 15, divisor = 3;
	int res = divide(dividend,divisor);
	cout<<res<<endl;

	cout<<INT_MAX<<INT_MIN<<endl;

	return 0;
}